compensations/src/main/java/org/jboss/narayana/compensations/impl/ParticipantImpl.java

```diff
@@ -92,9 +94,10 @@ public void close() throws WrongStateException, SystemException {
             ConfirmationHandler handler = instantiate(confirmationHandler);
             handler.confirm();
 
-            CompensationContext.close(currentTX);
             Thread.currentThread().setContextClassLoader(origClassLoader);
         }
+
+        CompensationContextLifetimeManager.getInstance().participantDestroyed(currentTX);
```

Comment:
These lifecycle methods can throw exceptions. I do not fully understand the semantics of such exceptions but you may need to put the  participantDestroyed calls inside a finally block.